### PR TITLE
Tethys4 upgrade

### PR DESCRIPTION
**Pull request by msouff**
Tethys 4 upgrade.

Most tests still have to be updated.

### CODE REVIEW

File: tethysext/atcore/public/js/form_processing.js [hunk 1]
```diff
@@ -28,11 +28,12 @@ function add_processing_div() {
 
 // Generate a GIF that covers page while loading
 function show_processing() {
-    $('#show-background').removeClass('hidden');
+    $('#show-background').removeClass('d-none');
 }
 
 function hide_processing() {
-    $('#show-background').addClass('hidden');
+    console.log()
```
Comment (swainn) [inline tethysext/atcore/public/js/form_processing.js:35]: Remove `console.log()`

File: tethysext/atcore/public/resource_workflows/dataset_workflow_results.css [hunk 2]
```diff
@@ -11,6 +11,16 @@
     margin-top: 20px;
 }
 
+table.dataTable > thead .sorting:before, table.dataTable > thead .sorting_asc:before, table.dataTable > thead .sorting_desc:before, table.dataTable > thead .sorting_asc_disabled:before, table.dataTable > thead .sorting_desc_disabled:before {
```
Comment (swainn) [inline tethysext/atcore/public/resource_workflows/dataset_workflow_results.css:14]: The convention is to add a new line after commas to make it more readable like so:

```css
table.dataTable > thead .sorting:before,
table.dataTable > thead .sorting_asc:before,
... {
    right: 1em;
}
```

File: tethysext/atcore/templates/atcore/map_view/map_view.html [hunk 3]
```diff
@@ -20,25 +20,27 @@
...
-             data-toggle="tab">{{ layer_tab_name }}</a>
+             data-bs-toggle="tab">{{ layer_tab_name }}</a>
         </li>
         {# LEGEND TAB #}
         {% if show_legends %}
           {% block legend_tab %}
             {% if legends %}
-            <li role="presentation" {% block disable_legend_tab %}{% endblock %}>
-              <a id="legend-tab-toggle"
+            <li class="nav-item" role="presentation" {% block disable_legend_tab %}{% endblock %}>
```
Comment (swainn) [inline tethysext/atcore/templates/atcore/map_view/map_view.html:36]: Move `{% block disable_legend_tab %}` to the `<a>` tag as above. The block is poorly named, but is used for the same purpose.

File: tethysext/atcore/templates/atcore/resource_workflows/components/workflow_tab.html [hunk 4]
```diff
@@ -4,11 +4,12 @@
 {% endcomment %}
 
 {% block workflow_tab %}
-<li role="presentation" class="active">
+<li role="presentation">
```
Comment (swainn) [inline tethysext/atcore/templates/atcore/resource_workflows/components/workflow_tab.html:7]: Should this have `class="nav-item"`?

File: tethysext/atcore/templates/atcore/resources/tabbed_resource_details.html [hunk 5]
```diff
@@ -24,7 +23,8 @@ <h1>{{ resource.name }}</h1>
     <a href="#{{ tab.slug }}"
```
Comment (swainn) [inline tethysext/atcore/templates/atcore/resources/tabbed_resource_details.html:23]: `<li>` above should have `nav-item` class, no?

File: tethysext/atcore/tests/integrated_tests/controllers/resources/tabbed_resource_details_tests.py [hunk 6]
```diff
@@ -64,7 +64,7 @@ def setUp(self):
         self.session.add(self.resource)
         self.session.commit()
 
-        self.app = mock.MagicMock(spec=TethysApp, namespace='app_namespace')
+        self.app = mock.MagicMock(spec=TethysApp, package='app_namespace')
```
Comment (swainn) [inline tethysext/atcore/tests/integrated_tests/controllers/resources/tabbed_resource_details_tests.py:67]: `package=` => `url_namespace=`

File: tethysext/atcore/tests/integrated_tests/controllers/resources/tabs/workflows_tab_tests.py [hunk 7]
```diff
@@ -78,7 +78,7 @@ def setUp(self):
         self.session.add(self.resource)
         self.session.commit()
 
-        self.app = mock.MagicMock(spec=TethysApp, namespace='foo_namespace')
+        self.app = mock.MagicMock(spec=TethysApp, package='foo_namespace')
```
Comment (swainn) [inline tethysext/atcore/tests/integrated_tests/controllers/resources/tabs/workflows_tab_tests.py:81]: `package=` => `url_namespace=`

File: tethysext/atcore/tests/utilities/sqlalchemy_helpers.py [hunk 8]
```diff
@@ -39,8 +39,7 @@ def setup_transaction_and_session(self):
         self.session = Session(connection)
 
     def tear_down_session_and_transaction(self):
-        self.session.close()
-        self.transaction.rollback()
+        self.session.close()  # Nested transactions seem to be deassociated on commit on 1.4
```
Comment (swainn) [inline tethysext/atcore/tests/utilities/sqlalchemy_helpers.py:42]: Was this comment meant to read "on close" rather then "on commit"?

File: tethysext/atcore/tests/integrated_tests/services/app_users/permissions_manager.py [hunk 9]
```diff
@@ -55,8 +55,8 @@ def setUp(self):
         self.staff_app_user = self.session.query(AppUser).filter(AppUser.username == AppUser.STAFF_USERNAME).one()
 
         # Permissions manager setup
-        self.namespace = 'foo'
-        self.apm = AppPermissionsManager(self.namespace)
+        self.package = 'foo'
```
Comment (swainn) [inline tethysext/atcore/tests/integrated_tests/services/app_users/permissions_manager.py:58]: `self.package	` => `self.url_namespace` (all instances)

File: tethysext/atcore/tests/integrated_tests/controllers/resource_workflows/workflow_view_tests/base_methods_tests.py [hunk 10]
```diff
@@ -84,7 +84,7 @@ def tearDown(self):
                 '.user_has_active_role', return_value=True)
     @mock.patch('tethys_apps.models.TethysApp')
     def test_get_context(self, mock_app, _, __):
-        mock_app.objects.get.return_value = mock.MagicMock(namespace='my_workspace')
+        mock_app.objects.get.return_value = mock.MagicMock(package='my_workspace')
```
Comment (swainn) [inline tethysext/atcore/tests/integrated_tests/controllers/resource_workflows/workflow_view_tests/base_methods_tests.py:87]: `package=` => `url_namespace=` (all instances)

File: tethysext/atcore/tests/integrated_tests/controllers/resource_workflows/workflow_results_view_tests.py [hunk 11]
```diff
@@ -102,7 +102,7 @@ def test_get_context(self, mock_get_context, mock_get_result, mock_validate, moc
     def test_get_result_url_name(self, mock_get_active_app):
         mock_request = mock.MagicMock()
         mock_workflow = mock.MagicMock(type='bar')
-        mock_get_active_app.return_value = mock.MagicMock(namespace='foo')
+        mock_get_active_app.return_value = mock.MagicMock(package='foo')
```
Comment (swainn) [inline tethysext/atcore/tests/integrated_tests/controllers/resource_workflows/workflow_results_view_tests.py:105]: `package=` => `url_namespace=`

File: tethysext/atcore/tests/integrated_tests/controllers/resource_workflows/resource_workflow_router_tests.py [hunk 12]
```diff
@@ -179,7 +183,7 @@ def test_get_with_step_id(self):
         )
 
         self.mock_redirect.assert_called()
-        url_name = f'{self.app.namespace}:generic_workflow_workflow_step_result'
+        url_name = f'{self.app.package}:generic_workflow_workflow_step_result'
```
Comment (swainn) [inline tethysext/atcore/tests/integrated_tests/controllers/resource_workflows/resource_workflow_router_tests.py:186]: `package` => `url_namespace` (all instances)

File: tethysext/atcore/tests/integrated_tests/controllers/resource_workflows/map_workflows/spatial_condor_job_mwv_tests.py [hunk 13]
```diff
@@ -194,7 +194,7 @@ def test_render_condor_jobs_table__read_only(self, mock_get_app, mock_get_active
         mock_get_app.return_value = app
 
         active_app = mock.MagicMock()
-        active_app.namespace = 'app_namespace'
+        active_app.package = 'app_namespace'
```
Comment (swainn) [inline tethysext/atcore/tests/integrated_tests/controllers/resource_workflows/map_workflows/spatial_condor_job_mwv_tests.py:197]: `package` => `url_namespace` (all instances)

File: tethysext/atcore/tests/integrated_tests/controllers/app_users/resource_status.py [hunk 14]
```diff
@@ -105,7 +105,7 @@ def test_default_back_url(self, mock_ga, mock_reverse):
         mock_request = self.request_factory.get('/foo/bar/status/?r={}'.format(self.resource_id))
 
         mock_app = mock.MagicMock()
-        mock_ga().namespace = 'test_namespace'
+        mock_ga().package = 'test_namespace'
```
Comment (swainn) [inline tethysext/atcore/tests/integrated_tests/controllers/app_users/resource_status.py:108]: `package` => `url_namespace` (all instances)

File: tethysext/atcore/tests/integrated_tests/controllers/app_users/modify_user.py [hunk 15]
```diff
@@ -110,7 +110,7 @@ def test__handle_modify_user_requests_post(self, _, mock_get_app_user_model, moc
 
         app_user.ROLES.get_no_organization_roles.return_value = ['Role1', 'Role2']
 
-        mock_get_active_app().namespace = 'NameSpace'
+        mock_get_active_app().package = 'NameSpace'
```
Comment (swainn) [inline tethysext/atcore/tests/integrated_tests/controllers/app_users/modify_user.py:113]: `package` => `url_namespace` (all instances)

File: tethysext/atcore/tests/integrated_tests/controllers/app_users/modify_resource_tests.py [hunk 16]
```diff
@@ -83,7 +83,7 @@ def setUp(self):
 
         get_active_app_patcher = mock.patch('tethysext.atcore.controllers.app_users.modify_resource.get_active_app')  # noqa: E501
         self.get_active_app = get_active_app_patcher.start()
-        self.get_active_app.return_value = mock.MagicMock(namespace='app_namespace')
+        self.get_active_app.return_value = mock.MagicMock(package='app_namespace')
```
Comment (swainn) [inline tethysext/atcore/tests/integrated_tests/controllers/app_users/modify_resource_tests.py:86]: `package` => `url_namespace` (all instances)

File: tethysext/atcore/tests/integrated_tests/controllers/app_users/modify_organization_tests.py [hunk 17]
```diff
@@ -75,7 +75,7 @@ def setUp(self):
 
         active_app_patcher = mock.patch('tethysext.atcore.controllers.app_users.modify_organization.get_active_app')
         self.mock_get_active_app = active_app_patcher.start()
-        self.mock_get_active_app.return_value = mock.MagicMock(namespace='app_namespace')
+        self.mock_get_active_app.return_value = mock.MagicMock(package='app_namespace')
```
Comment (swainn) [inline tethysext/atcore/tests/integrated_tests/controllers/app_users/modify_organization_tests.py:78]: `package` => `url_namespace` (all instances)

File: tethysext/atcore/tests/integrated_tests/controllers/app_users/mixins.py [hunk 18]
```diff
@@ -103,7 +103,7 @@ def tearDown(self):
     def test_default_back_url(self, mock_reverse, mock_aa):
         mock_request = self.request_factory.get('/foo/bar/')
         mock_resource_id = self.resource_id
-        mock_aa.return_value = mock.MagicMock(namespace='test1')
+        mock_aa.return_value = mock.MagicMock(package='test1')
```
Comment (swainn) [inline tethysext/atcore/tests/integrated_tests/controllers/app_users/mixins.py:106]: `package` => `url_namespace` (all instances)

File: tethysext/atcore/tests/integrated_tests/controllers/app_users/manage_resources.py [hunk 19]
```diff
@@ -345,7 +345,7 @@ def test_get_resource_action_error(self, mock_reverse):
         ret = manage_resources.get_resource_action('', '', '', mock_resource)
 
         mock_reverse.assert_called_with(
-            f'{manage_resources._app.namespace}:{mock_resource.SLUG}_resource_details',
+            f'{manage_resources._app.package}:{mock_resource.SLUG}_resource_details',
```
Comment (swainn) [inline tethysext/atcore/tests/integrated_tests/controllers/app_users/manage_resources.py:348]: `package` => `url_namespace` (all instances)

File: tethysext/atcore/tests/integrated_tests/controllers/app_users/add_existing_user.py [hunk 20]
```diff
@@ -93,7 +93,7 @@ def test__handle_modify_user_requests(self, mock_get_app_usermodel, _,
         mock_no_of_orgs = mock.MagicMock()
         mock_get_app_usermodel().ROLES.get_no_organization_roles.return_value = mock_no_of_orgs
 
-        mock_get_active_app().namespace = 'NameSpace'
+        mock_get_active_app().package = 'NameSpace'
```
Comment (swainn) [inline tethysext/atcore/tests/integrated_tests/controllers/app_users/add_existing_user.py:96]: `package` => `url_namespace` (all instances)

File: tethysext/atcore/templates/atcore/app_users/modify_resource.html [hunk 21]
```diff
@@ -103,6 +103,7 @@ <h1>{% if editing %}Edit{% else %}New{% endif %} {{ type_singular }}</h1>
   <link href="{% static 'atcore/file_input/css/file_input.css' %}" rel="stylesheet"/>
   {% endif %}
   <link href="{% static 'atcore/css/form_processing.css' %}" rel="stylesheet"/>
+  <link href="{% static 'atcore/css/center.css' %}" rel="stylesheet"/>
```
Comment (swainn) [inline tethysext/atcore/templates/atcore/app_users/modify_resource.html:106]: Is center.css needed here? I don't see any of the classes that are used to center things being used in the changes.

File: tethysext/atcore/public/css/center.css [hunk 22]
```diff
@@ -6,7 +6,7 @@
 .center-parent:before {
   content: '';
   display: inline-block;
-  height: 100%;
+  height: 50%;
```
Comment (swainn) [inline tethysext/atcore/public/css/center.css:9]: Why was this change needed? This should be 100%.

File: tethysext/atcore/services/geoserver_api.py [hunk 23]
```diff
@@ -311,7 +311,9 @@ def create_postgis_store(self, workspace, name, db_host, db_port, db_name, db_us
         )
 
         # Return with error if this doesn't work
-        if response.status_code != 201:
```
Comment (swainn) [inline tethysext/atcore/services/geoserver_api.py:314]: What would it take to remove this file entirely and replace it's usage with the equivalent methods in the tethys_dataset_services API?

File: tethysext/atcore/public/js/messages.js [hunk 24]
```diff
@@ -1,7 +1,6 @@
 function show_message_box(id, level, message) {
     let message_box = '<div id="' + id + '" class="alert alert-' + level + ' alert-dismissible" role="alert">'
-                    +   '<button type="button" class="close" data-dismiss="alert">'
-                    +     '<span aria-hidden="true">&times;</span>'
+                    +   '<button type="button" class="btn-close" data-bs-dismiss="alert">'
                     +     '<span class="sr-only">Close</span>'
```
Comment (swainn) [inline tethysext/atcore/public/js/messages.js:4]: Is the `<span>` needed now?

**Review by swainn — CHANGES_REQUESTED** [review]
See comments.
Also, how many tests are broken now?

File: tethysext/atcore/controllers/resource_workflows/workflow_view.py [hunk 25]
```diff
@@ -313,7 +313,7 @@ def get_workflow_url_name(request, workflow):
             str: name of the url pattern for the given workflow views.
         """
         active_app = get_active_app(request)
-        url_map_name = '{}:{}_workflow'.format(active_app.namespace, workflow.type)
+        url_map_name = '{}:{}_workflow'.format(active_app.package, workflow.type)
```
[the remark this answers is not in the document]
Comment (msouff) [reply tethysext/atcore/controllers/resource_workflows/workflow_view.py:316]: This one shouldn't need change since it's using the url name from the url mapping instead of the url itself.

File: tethysext/atcore/controllers/resource_workflows/workflow_results_view.py [hunk 26]
```diff
@@ -86,7 +86,7 @@ def get_result_url_name(request, workflow):
             str: name of the url pattern for the given workflow step views.
         """
         active_app = get_active_app(request)
-        url_map_name = '{}:{}_workflow_step_result'.format(active_app.namespace, workflow.type)
+        url_map_name = '{}:{}_workflow_step_result'.format(active_app.package, workflow.type)
```
[the remark this answers is not in the document]
Comment (msouff) [reply tethysext/atcore/controllers/resource_workflows/workflow_results_view.py:89]: See previous comment

File: tethysext/atcore/controllers/resources/tabs/workflows_tab.py [hunk 27]
```diff
@@ -134,7 +134,7 @@ def get_context(self, request, session, resource, context, *args, **kwargs):
 
         for workflow in workflows:
             status = workflow.get_status()
-            app_namespace = self.get_app().namespace
+            app_namespace = self.get_app().package
```
[the remark this answers is not in the document]
Comment (msouff) [reply tethysext/atcore/controllers/resources/tabs/workflows_tab.py:137]: See https://github.com/Aquaveo/tethysext-atcore/pull/63#discussion_r957749604

File: tethysext/atcore/models/types/guid.py [hunk 28]
```diff
@@ -11,6 +11,7 @@ class GUID(TypeDecorator):
     CHAR(32), storing as stringified hex values.
     """
     impl = CHAR
+    cache_ok = True
```
[the remark this answers is not in the document]
Comment (msouff) [reply tethysext/atcore/models/types/guid.py:14]: I think we should be okay changing it to False. The issue was with the default value of None.

File: tethysext/atcore/tests/utilities/sqlalchemy_helpers.py (hunk 8 above)
Comment (msouff) [reply tethysext/atcore/tests/utilities/sqlalchemy_helpers.py:42]: See https://github.com/sqlalchemy/sqlalchemy/issues/6155

We could add the "if is_active" check before we try to rollback but for now I'd be okay to just add it back since the tests need further revision anyway.

File: tethysext/atcore/tests/integrated_tests/controllers/resource_workflows/resource_workflow_router_tests.py (hunk 12 above)
Comment (msouff) [reply tethysext/atcore/tests/integrated_tests/controllers/resource_workflows/resource_workflow_router_tests.py:186]: See https://github.com/Aquaveo/tethysext-atcore/pull/63#discussion_r957749604

File: tethysext/atcore/templates/atcore/app_users/modify_resource.html (hunk 21 above)
Comment (msouff) [reply tethysext/atcore/templates/atcore/app_users/modify_resource.html:106]: You are right. that was not being used so I removed it. I also found a resize bug in the spatial reference droupdown of this template and took care of it.

Comment (msouff) [reply tethysext/atcore/templates/atcore/app_users/modify_resource.html:106]: Actually, center.css is used to display the loading bar so I've put it back. See comment below for more info.

File: tethysext/atcore/public/css/center.css (hunk 22 above)
Comment (msouff) [reply tethysext/atcore/public/css/center.css:9]: At first I couldn't remember why we changed this, but this is what displays the loading bar when we add a resource. We have this ::before so that we can center the loading bar. If the height is 100% you can't see the bar, just a gray screen. 50% for the ::before makes it so the bar shows right after, near the vertical center.

File: tethysext/atcore/services/geoserver_api.py (hunk 23 above)
Comment (msouff) [reply tethysext/atcore/services/geoserver_api.py:314]: Four hours at most. I checked and GeoServerAPI is only used in the init command and the base spatial manager. All uses call the class so it would be easy to just switch the import to come from tds. The rest would be cleaning up, including a bunch of tests that could be removed.

File: tethysext/atcore/public/js/messages.js (hunk 24 above)
Comment (msouff) [reply tethysext/atcore/public/js/messages.js:4]: This is not needed anymore. Removed.

File: tethysext/atcore/services/geoserver_api.py (hunk 23 above)
Comment (swainn) [reply tethysext/atcore/services/geoserver_api.py:314]: Do it.

File: tethysext/atcore/public/js/form_processing.js (hunk 1 above)
Comment (swainn) [reply tethysext/atcore/public/js/form_processing.js:35]: @msouff miss this?

**Review by swainn — COMMENTED** [review]
Responded to comments

File: tethysext/atcore/tests/utilities/sqlalchemy_helpers.py (hunk 8 above)
Comment (swainn) [reply tethysext/atcore/tests/utilities/sqlalchemy_helpers.py]: So the reason for this change is that you were getting lots of warnings when running tests?

Comment (msouff) [reply tethysext/atcore/tests/utilities/sqlalchemy_helpers.py]: Yes, lots of them.

File: tethysext/atcore/services/geoserver_api.py (hunk 23 above)
Comment (msouff) [reply tethysext/atcore/services/geoserver_api.py:314]: Yes, agwa would require some refactoring as well. It uses atcore base spatial manager class and calls various gs_api functions using "self.gs_api"

These are the functions agwa uses: get_layer_extent, create_layer, create_style, delete_style. The main change in these is that before we took workspace and name parameters and now we take a store_id or layer_id that is a combination of the two (e.g. workspace:name), so the refactoring would not be too complicated.

**Review by swainn — COMMENTED** [review]
Will work with you on tests soon.

**Review by swainn — APPROVED** [review]
